### PR TITLE
[Google Blockly] Convert remaining* addons to TypeScript

### PR DESCRIPTION
**Pull request by molly-moen**
With this PR every file directly in the `blockly/addons` file will be converted to TypeScript 🎉 I did not convert the files under `blockly/addons/plusMinusBlocks` because they are mostly copied from a Blockly plugin that has not been converted to TypeScript. For now I'm not prioritizing converting these.

These changes were pretty straightforward, mostly types and null checks. I added comments where things were more interesting. Of note is I was able to delete `cdoTouch` because it was overriding a defunct method.
The files converted are:
- `cdoSpritePointer`
- `cdoTrashcan`
- `cdoVariables`
- `cdoVerticalFlyout`
- `contextMenu`
- `functionEditorConstants`
- `unknownBlock`

## Links
- jira ticket: [CT-335](https://codedotorg.atlassian.net/browse/CT-335)

## Testing story
I tested the affected functionality locally.

## Follow-up work
Here is what is left under `blockly` and whether I think it's worth converting:
- (folder) `/addons/plusMinusBlocks`: No, it's mostly copied over from Blockly and is not in TypeScript on their end. If they convert `@blockly/block-plus-minus` to TypeScript we could pull that in.
- (folder) `/customBlocks/cdoBlockly`: No, I don't think it's worth our time to convert cdo Blockly code
- (file) `cdoBlocklyWrapper`: No, also cdo Blockly code
- (folder) `/customBlocks/googleBlockly`: Yes. There are 15 files in here of varying sizes. It's a non-zero amount of work but _hopefully_ the work to convert everything else will make these straightforward.
- (folder) `/themes`: Maybe. These are mostly constants, and we are doing some weird things here with module files to convert styles to accessible styles. We could convert the theme files and not the constant files. Overall these are lower priority.

TL;DR, the Google Blockly custom block files are what we should convert next, and we can decide at that point if we want to do any of the remaining theme files.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoTouch.js [hunk 1]
```diff
@@ -1,15 +0,0 @@
-export default function initializeTouch(blocklyWrapper) {
-  // Aliasing Google's setClientFromTouch() so that we can override it, but still be able
-  // to call Google's setClientFromTouch() in the override function.
-  blocklyWrapper.Touch.originalSetClientFromTouch =
-    blocklyWrapper.Touch.setClientFromTouch;
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoTouch.js:5]: This function does not exist on `Blockly.Touch` anymore. We added this override to fix flappy tests. I verified that they pass without this change, which makes sense as since this function doesn't exist, overriding it does nothing.

File: apps/src/blockly/addons/cdoSpritePointer.ts [hunk 2]
```diff
@@ -68,53 +75,72 @@ export function updatePointerBlockImage(
...
+  imageSourceBlock: ExtendedBlockSvg,
+  imageIndexOnSource: number
+) {
   const targetConnection =
     imageSourceBlock.inputList[imageIndexOnSource]?.connection
       ?.targetConnection;
   // We only want to get the image from a connection block that is not an insertion marker.
   // If the block is an insertion marker that means the block is being dragged
   // and is not yet connected to the image source block.
+  const sourceBlock = targetConnection?.getSourceBlock();
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoSpritePointer.ts:93]: `sourceBlock_` is private, so we use `getSourceBlock()` instead (which returns `sourceBlock_`)

File: apps/src/blockly/addons/cdoSpritePointer.ts [hunk 3]
```diff
@@ -68,53 +75,72 @@ export function updatePointerBlockImage(
...
-function changePointerImage(imageUrl, block) {
+function changePointerImage(
+  imageUrl: string | undefined,
+  block: ExtendedBlockSvg
+) {
   if (!imageUrl || imageUrl.length === 0) {
     resetPointerImageToLongString(block);
     return;
   }
-  const textInput = block.inputList[0].fieldRow[0];
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoSpritePointer.ts:108]: almost the exact same code was used here and in `resetPointerImageToLongString`. I pulled it into a helper method `updatePointerImageHelper`.

File: apps/src/blockly/addons/cdoTrashcan.ts [hunk 4]
```diff
@@ -360,13 +397,11 @@ export default class CdoTrashcan extends GoogleBlockly.DeleteArea {
    * @return {boolean} Whether the block or bubble provided should be returned to
    *     drag start.
    */
-  shouldPreventMove(_dragElement) {
+  shouldPreventMove(_dragElement: IDraggable) {
     return false;
   }
 }
 
-CdoTrashcan.TRASH_URL = '/blockly/media/trash.png';
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoTrashcan.ts:368]: I verified that this was not used anywhere. It was originally used in this class in [this PR](https://github.com/code-dot-org/code-dot-org/pull/37002/files), but as some point we moved to use `this.TRASH_URL`.

File: apps/src/blockly/addons/cdoVerticalFlyout.ts [hunk 5]
```diff
@@ -49,22 +52,22 @@ export default class VerticalFlyout extends GoogleBlockly.VerticalFlyout {
...
+          const oldX = block.getRelativeToSurfaceXY().x;
+          let newX = flyoutWidth / this.workspace_.scale - this.MARGIN;
           if (!block.outputConnection) {
             newX -= this.tabWidth_;
           }
           block.moveBy(newX - oldX, 0);
         }
-        if (block.flyoutRect_) {
-          this.moveRectToBlock_(block.flyoutRect_, block);
+        if (this.rectMap_.has(block)) {
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoVerticalFlyout.ts:62]: This change comes from the [original blockly function](https://github.com/google/blockly/blob/71185b5582401cd499f0e4a8ed444e7c4d531580/core/flyout_vertical.ts#L361-L363). It has been updated since we copied it over. `block.flyoutRect_` does not exist anymore.

File: apps/src/blockly/addons/cdoSpritePointer.ts [hunk 6]
```diff
@@ -68,53 +75,72 @@ export function updatePointerBlockImage(
...
     // Blocks of type gamelab_allSpritesWithAnimation have an input with one field (the costume
     // picker dropdown).
-    return (
-      targetConnection.sourceBlock_.inputList[0].fieldRow[0].imageElement.getAttribute(
-        'xlink:href'
-      ) || ''
-    );
+    // imageElement is a private property on FieldDropdown.
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const imageElement = (sourceBlock.inputList[0].fieldRow[0] as any)
```
Comment (mikeharv) [inline apps/src/blockly/addons/cdoSpritePointer.ts]: Unrelated to the TS changes, but I think we could replace `inputList[0].fieldRow[0]` with `getField('ANIMATION')`. Would that feel cleaner and/or easier to read?

image

I don't expect this block to change much, but if, for example, we ever added a label before the field ("any [costume]" or something), then this getting the field we want by name would be more reliable.

File: apps/src/blockly/addons/cdoTouch.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoTouch.js:5]: There was a significant update a few Blockly versions back that changed how touch/mouse/pointer events. In fact, I believe Blockly always uses Pointer now, so it makes sense that this might not be doing anything.

File: apps/src/blockly/addons/cdoTrashcan.ts [hunk 7]
```diff
@@ -159,39 +174,50 @@ export default class CdoTrashcan extends GoogleBlockly.DeleteArea {
...
     if (blocklyEvent.type === Blockly.Events.BLOCK_DRAG) {
+      const event = blocklyEvent as BlockDrag;
       let trashcanVisibility = 'hidden';
       let toolboxVisibility = 'visible';
       // Don't show the trashcan if the block is being dragged out of the toolbox.
-      const isDraggingFromToolbox = !!this.workspace.currentGesture_?.flyout_;
-      if (!isDraggingFromToolbox && blocklyEvent.isStart) {
+      // This property is private, but this is the only way to find out this information.
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const isDraggingFromToolbox = !!(this.workspace.currentGesture_ as any)
```
Comment (mikeharv) [inline apps/src/blockly/addons/cdoTrashcan.ts:185]: It looks like we might be able to use `getGesture` here: 
image

https://github.com/google/blockly/blob/71185b5582401cd499f0e4a8ed444e7c4d531580/core/workspace_svg.ts#L2471-L2503

File: apps/src/blockly/addons/cdoTrashcan.ts [hunk 8]
```diff
@@ -159,39 +174,50 @@ export default class CdoTrashcan extends GoogleBlockly.DeleteArea {
...
 
       // Shadow blocks can/should be successfully deleted
       // when dragged in conjunction with another deletable block;
       // however, isDeletable() returns false for shadow blocks,
       // so we manually override here.
-      const isDeletable = blocklyEvent.blocks.every(
+      const isDeletable = event.blocks?.every(
         block => block.isDeletable() || block.isShadow()
       );
+      if (!this.notAllowed_) {
```
Comment (mikeharv) [inline apps/src/blockly/addons/cdoTrashcan.ts:218]: The `notAllowed_` svg is created during `init()` -> `this.createTrashcanSvg()`, and this `workspaceChangeHandler` method is added a change listener a few lines earlier in `init()`. Is it possible that changing the order of these two thing in `init()` would help us to guarantee that the SVG exists at this point?

File: apps/src/blockly/addons/cdoTrashcan.ts (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoTrashcan.ts:368]: Confirmed that this looks redundant and unused now.
https://github.com/code-dot-org/code-dot-org/commit/b41904699ebd3a321e1cbb9ac419292afd30c778

**Review by mikeharv — APPROVED** [review]
A couple suggestions and questions but no blockers!

File: apps/src/blockly/addons/cdoTrashcan.ts (hunk 7 above)
Comment (molly-moen) [reply apps/src/blockly/addons/cdoTrashcan.ts:185]: `currentGesture_` is actually a public field, it's `flyout` that is the private field. `getGesture()` can [change/create](https://github.com/google/blockly/blob/71185b5582401cd499f0e4a8ed444e7c4d531580/core/workspace_svg.ts#L2480) `currentGesture_` and I'm not sure that's what we would want.

File: apps/src/blockly/addons/cdoTrashcan.ts (hunk 8 above)
Comment (molly-moen) [reply apps/src/blockly/addons/cdoTrashcan.ts:218]: I don't think TypeScript is smart enough for that. It will only guarantee things aren't null if we initialize them in the constructor.

File: apps/src/blockly/addons/cdoSpritePointer.ts (hunk 6 above)
Comment (molly-moen) [reply apps/src/blockly/addons/cdoSpritePointer.ts]: updated!
